src/toplevel_generators.jl

```diff
@@ -0,0 +1,182 @@
+            @static if isdir(joinpath(dirname($(pkg_dir)), "override"))
+                return joinpath(dirname($(pkg_dir)), "override")
+            else
+                # We explicitly use `macrocall` here so that we can manually pass the `__source__`
+                # argument, to avoid `@artifact_str` trying to lookup `Artifacts.toml` here.
+                return $(Expr(:macrocall, Symbol("@artifact_str"), __source__, src_name))
+            end
+        end
+
+        # We sub off to JLLWrappers' dev_jll, but avoid backedges
+        dev_jll() = Base.invokelatest(JLLWrappers.dev_jll, $(src_name))
+    end
+end
+
+
+"""
+    generate_wrapper_load(src_name, pkg_uuid)
+
+Because each platform could have completely disjoint products, we embed the information
+for each in its own `.jl` file, named by triplet, and stored within the `src/wrappers`
+directory.  This method generates the code to load `Artifacts.toml`, parse it for
+artifacts, find the one that matches the host platform, then load the matching wrapper.
+"""
+function generate_wrapper_load(src_name, pkg_uuid, __source__)
+    jll_name = "$(src_name)_jll"
+    pkg_dir = dirname(String(__source__.file))
+    return quote
+        global best_wrapper
+        # Load Artifacts.toml file and select best platform at compile-time, since this is
+        # running at toplevel, and therefore will be run completely at compile-time.  We use
+        # a `let` block here to avoid storing unnecessary data in our `.ji` files
+        best_wrapper = let
+            artifacts_toml = joinpath($(pkg_dir), "..", "Artifacts.toml")
+            valid_wrappers = Dict{Platform,String}()
+            artifacts = load_artifacts_toml(artifacts_toml; pkg_uuid=$(pkg_uuid))[$(src_name)]
+            # If it's a Dict, that means this is an AnyPlatform artifact, act accordingly:
+            if isa(artifacts, Dict)
+                joinpath($(pkg_dir), "wrappers", "any.jl")
+            else
+                # Otherwise, it's a Vector, and we must select the best platform
+                for e in artifacts
+                    platform = unpack_platform(e, $(jll_name), artifacts_toml)
+
+                    # Filter platforms based on what wrappers we've generated on-disk
+                    wrapper_file = joinpath($(pkg_dir), "wrappers", string(triplet(platform), ".jl"))
+                    if isfile(wrapper_file)
+                        valid_wrappers[platform] = wrapper_file
+                    end
+                end
+
+                # From the available options, choose the best wrapper script
+                select_platform(valid_wrappers)
+            end
+        end
+
+        # Load in the wrapper, if it's not `nothing`!
+        if best_wrapper === nothing
+            @debug(string("Unable to load ", $(src_name), "; unsupported platform ", triplet(HostPlatform())))
+            is_available() = false
```

Comment:
Wow, this is a much cleaner way to define this function than what we're doing now!

Reply:
:)